syncer/filter_test.go

```diff
@@ -91,41 +71,172 @@ func (s *testFilterSuite) TestSkipQueryEvent(c *C) {
 		tables        []*filter.Table
 		expectSkipped bool
 	}{
-		// test global rule
 		{
-			"drop table tx.test",
-			[]*filter.Table{{Schema: "tx", Name: "test"}},
+			// system table
+			"create table mysql.test (id int)",
+			[]*filter.Table{{Schema: "mysql", Name: "test"}},
+			true,
+		}, {
+			// test filter one event
+			"drop table foo.test",
+			[]*filter.Table{{Schema: "foo", Name: "test"}},
+			false,
+		}, {
+			"create table foo.test (id int)",
+			[]*filter.Table{{Schema: "foo", Name: "test"}},
 			true,
 		},
+	}
+	p := parser.New()
+	for _, ca := range cases {
+		stmt, err := p.ParseOneStmt(ca.sql, "", "")
+		c.Assert(err, IsNil)
+		skipped, err2 := syncer.skipQueryEvent(ca.tables, stmt, ca.sql)
+		c.Assert(err2, IsNil)
+		c.Assert(skipped, Equals, ca.expectSkipped)
+	}
+}
+
+func (s *testFilterSuite) TestSkipRowsEvent(c *C) {
+	syncer := &Syncer{}
+	filterRules := []*bf.BinlogEventRule{
 		{
-			"create table tx.test (id int)",
-			[]*filter.Table{{Schema: "tx", Name: "test"}},
-			false,
+			SchemaPattern: "foo*",
+			TablePattern:  "",
+			Events:        []bf.EventType{bf.InsertEvent},
+			SQLPattern:    []string{""},
+			Action:        bf.Ignore,
 		},
-		// test schema rule
+	}
+	var err error
+	syncer.binlogFilter, err = bf.NewBinlogEvent(false, filterRules)
+	c.Assert(err, IsNil)
+	syncer.onlineDDL = mockOnlinePlugin{}
+
+	cases := []struct {
+		table     *filter.Table
+		eventType replication.EventType
+		expected  bool
+	}{
 		{
-			"create table foo.test(id int)",
-			[]*filter.Table{{Schema: "foo", Name: "test"}},
+			// test un-realTable
+			&filter.Table{Schema: "foo", Name: "_test_gho"},
+			replication.UNKNOWN_EVENT,
+			true,
+		}, {
+			// test filter one event
+			&filter.Table{Schema: "foo", Name: "test"},
+			replication.WRITE_ROWS_EVENTv0,
+			true,
+		}, {
+			&filter.Table{Schema: "foo", Name: "test"},
+			replication.UPDATE_ROWS_EVENTv0,
+			false,
+		}, {
+			&filter.Table{Schema: "foo", Name: "test"},
+			replication.DELETE_ROWS_EVENTv0,
 			false,
 		},
+	}
+	for _, ca := range cases {
+		needSkip, err2 := syncer.skipRowsEvent(ca.table, ca.eventType)
+		c.Assert(err2, IsNil)
+		c.Assert(needSkip, Equals, ca.expected)
+	}
+}
+
+func (s *testFilterSuite) TestFilterOneEvent(c *C) {
+	cfg := &config.SubTaskConfig{
+		BAList: &filter.Rules{
+			IgnoreDBs: []string{"s1"},
+		},
+	}
+	syncer := NewSyncer(cfg, nil)
+	var err error
+	syncer.baList, err = filter.New(syncer.cfg.CaseSensitive, syncer.cfg.BAList)
+	c.Assert(err, IsNil)
+	// test binlog filter
+	filterRules := []*bf.BinlogEventRule{
 		{
-			"rename table foo.test to foo.test1",
```

Comment:
test a RENAME SQL because it has more than one table name

Reply:
I agree. `RENAME` SQL has more than one table name, but L113 only has one table name...

Reply:
the logic is RENAME will be skipped when any table is skipped (`if len(tbs) != len(tables)`)

Reply:
I know the logic. I mean the input `tables` is from `FetchDDLTables`. If SQL is `RENAME`, the input `Tables` will contain [old1, new1], but the test case in L113 is only [old1]. So the former test maybe have a problem.